### DISCUSSION TITLE
Cost function value suddenly spikes when using gradient mpc and cuda graphs enabled

### DISCUSSION THREAD

**Discussion by akselva** [NVlabs/curobo#165] [Software Q&A]
Hi, I'm posting this under discussion because (as far as i know) gradient based mpc is not a core part of curobo development.

I think curobo is awesome, and I plan on integrating curobo mpc (particle, gradient based or a combination) as part of my research. I am trying to get the gradient-based version to work properly, however I observe some strange behavior when cuda graphs are enabled. For the first few mpc iterations, both mpc with and without cuda graph give identical results, but after about 7 iterations the cg_enabled version computes extremely high cost values for the rollouts. I am still debugging the issue, and I would really appreciate if you could take a quick look at this.

This is a typical result when running the mpc_example.py. I have set inner_iters = n_iters = 25 (for the sake of debugging), but the same issue seems to happen as long as inner_iters > 1.

```
iteration  0
best cost cg:  tensor(790.749)  , best cost:  tensor(790.749)
iteration  1
best cost cg:  tensor(739.112)  , best cost:  tensor(739.112)
iteration  2
best cost cg:  tensor(702.419)  , best cost:  tensor(702.419)
iteration  3
best cost cg:  tensor(656.656)  , best cost:  tensor(656.656)
iteration  4
best cost cg:  tensor(607.618)  , best cost:  tensor(607.618)
iteration  5
best cost cg:  tensor(577.563)  , best cost:  tensor(577.563)
iteration  6
best cost cg:  tensor(531.463)  , best cost:  tensor(531.463)
iteration  7
best cost cg:  tensor(2962512.250)  , best cost:  tensor(495.518)
iteration  8
best cost cg:  tensor(414225.156)  , best cost:  tensor(468.728)
```

My current best guess is that the issue is related to a potential cuda graph pool memory rewrite/corruption during updation of goal buffers, however I am struggling to really trace what is happening in that part of the code.

Here is a script to reproduce the error and compare cg vs no_cg, based on the mpc_example script. It must be placed in the examples folder
Set n_iters = inner_iters = 25 to reproduce the above output
I have also increased inital best cost value to 5.000.000 in newton_base -> reset()
[grad_mpc_cg_error_reproduce.zip](https://github.com/NVlabs/curobo/files/14405054/grad_mpc_cg_error_reproduce.zip)

**Comment by balakumar-s** [comment]
Thanks for trying this out, we can check on our end on what's happening. 

We also have a reactive motion_gen mode which runs trajectory optimization from non-zero velocities. This is similar to MPC but uses IK for initialization and there is no warm start. You can try this mode by running our isaac sim example: `omni_python examples/isaac_sim/motion_gen_reacher.py --reactive`

**Reply by akselva** [reply]
Thank you! Will check out the reactive motion gen.

**Reply by YoruCathy** [reply]
Hi all, I am having a similar issue with my project. Did you manage to get this working?

**Reply by akselva** [reply]
It is a long time since I worked on it, but as I recall I think I managed to solve the issue somewhat. I am a bit busy at the moment, but I will take a look at the changes I made to the code and get back to you as soon as I can

**Reply by YoruCathy** [reply]
Thanks so much! I'm actually trying to get this working for CoRL, so I would very appreciate it if you could share the code soon. Thanks again!

**Reply by balakumar-s** [reply]
We pushed a bug fix for gradients in the latest commit. This could resolve your issue @YoruCathy.
